src/election.ts

```diff
@@ -0,0 +1,282 @@
+import BigNumber from 'bignumber.js';
+import * as EventEmitter from 'events';
+import { EtcdNoLeaderError, EtcdNotLeaderError } from './errors';
+import { Lease } from './lease';
+import { Namespace } from './namespace';
+
+/**
+ * Implmentation of etcd election.
+ * @see https://github.com/coreos/etcd/blob/master/clientv3/concurrency/election.go
+ *
+ * @example
+ * const client = new Etcd3()
+ * const election = new Election(client, 'singleton_service')
+ * const id = BigNumber.random().toString()
+ *
+ * // process will hang here until elected
+ * await election.campaign(id)
+ */
+export class Election extends EventEmitter {
+  public static readonly prefix = 'election';
+
+  private readonly namespace: Namespace;
+  private lease: Lease;
+
+  private leaseId = '';
+  private _leaderKey = '';
+  private _leaderRevision = '';
+  private _isCampaigning = false;
+  private _isObserving = false;
+
+  public get leaderKey(): string { return this._leaderKey; }
+  public get leaderRevision(): string { return this._leaderRevision; }
+  public get isReady(): boolean { return this.leaseId.length > 0; }
+  public get isCampaigning(): boolean { return this._isCampaigning; }
+  public get isObserving(): boolean { return this._isObserving; }
+
+  constructor(public readonly parent: Namespace,
+              public readonly name: string,
+              public readonly ttl: number = 60) {
+    super();
+    this.namespace = parent.namespace(this.getPrefix());
+    this.lease = this.namespace.lease(ttl);
+    this.lease.on('lost', () => this.onLeaseLost());
+  }
+
+  public on(event: 'leader', listener: (leaderKey: string) => void): this;
+  public on(event: 'error', listener: (error: any) => void): this;
+  public on(event: string|symbol, listener: Function): this;
+  public on(event: string|symbol, listener: Function): this {
+    super.on(event, listener);
+    if (this.shouldObserve(event)) {
```

Comment:
You probably want to use https://nodejs.org/api/events.html#events_event_newlistener instead, less hacky.

Reply:
Cool! It did not appear in declaration, thanks!.